perf(embedding): default embedding creation to base64

manekinekko opened the pull request:
Requesting base64 encoded embeddings returns smaller body sizes, on average ~60% smaller than float32 encoded. In other words, the size of the response body containing embeddings in float32 is ~2.3x bigger than base64 encoded embedding.

Closes #1310

<!-- Thank you for contributing to this project! -->
<!-- The code in this repository is all auto-generated, and is not meant to be edited manually. -->
<!-- We recommend opening an Issue instead, but you are still welcome to open a PR to share for -->
<!-- an improvement if you wish, just note that we are unlikely to merge it as-is. -->

- [x] I understand that this repository is auto-generated and my pull request may not be merged

## Changes being requested

We always request embedding creating encoded as base64, and then decoded them to float32 based on the user's provided encoding_format parameter.

## Additional context & links

After running a few [benchmarks](https://github.com/manekinekko/rich-bench-node), requesting base64 encoded embeddings returns smaller body sizes, on average **~60% smaller** than float32 encoded. In other words, the size of the response body containing embeddings in float32 is **~2.3x bigger** than base64 encoded embedding.

This performance improvement could translate to:
- ✅ Faster HTTP responses
- ✅ Less bandwidth used when generating multiple embeddings

This is the result of a request that creates embedding from a 10kb chunk, run 10 times (the number are the size of response body in kb):

| Benchmark        | Min (ms) | Max (ms)  | Mean (ms)  | Min (+)          | Max (+)           | Mean (+)          |
|-----------------|---------:|----------:|----------:|-----------------:|-------------------:|------------------:|
| float32 vs base64 |   41.742 | 19616.000 |  9848.819 | 40.094 (3.9%)    | 8351.000 (57.4%)  | 4206.126 (57.3%)  |

Read more #1310

RobertCraigie commented on src/resources/embeddings.ts:
```diff
@@ -9,9 +9,46 @@ export class Embeddings extends APIResource {
+    const base64Response = this._client.post<EmbeddingCreateParams, CreateEmbeddingResponse>('/embeddings', {
+      body: {
+        ...body,
+        // Force base64 encoding for vector embeddings creation
+        // See https://github.com/openai/openai-node/issues/1310
+        encoding_format: 'base64',
```
I don't think we want to *always* use `base64`, if the user explicitly asked for a different format we should have the exact same behaviour as we do prior to this PR which is to just let them.

RobertCraigie replied:
yes I think the tradeoff of not getting the perf optimisation is worth it. in general we do not want to get in the users way, if the user explicitly requests `float` then we should just send `float`.

having a way to opt-out of certain SDK things can be very helpful, especially if say, a new encoding format is added, then users wouldn't be able to actually use that until we update this logic to support it. however if we just send custom formats as-is then it'll just work, no matter what SDK version they're on

manekinekko replied:
@RobertCraigie I updated the logic. We now default to base64 only if the user does not provide en encoding_format.

manekinekko replied:
@RobertCraigie I updated the logic based on your feedback. We now default to base64 _only_ if the user does not provide en encoding_format. Let me know if that works for you.

manekinekko replied:
@RobertCraigie Please let me know if there is anything else missing.

RobertCraigie commented on src/resources/embeddings.ts:
```diff
@@ -9,9 +9,46 @@ export class Embeddings extends APIResource {
+            console.log(embeddingBase64Obj);
+            const embeddingBase64Str = embeddingBase64Obj.embedding as unknown as string;
+            embeddingBase64Obj.embedding = Array.from(
+              new Float32Array(Buffer.from(embeddingBase64Str, 'base64').buffer),
```
`Buffer` is a Node.js specific API, we need to use something that is available everywhere, would be great to add a generic helper function in `core.ts`.

manekinekko replied:
Sure. I will send a fix.

manekinekko replied:
Moved logic to core.ts

RobertCraigie commented on src/resources/embeddings.ts:
```diff
@@ -9,9 +9,46 @@ export class Embeddings extends APIResource {
+      return base64Response._thenUnwrap((response) => {
+        if (response && response.data) {
+          response.data.forEach((embeddingBase64Obj) => {
+            console.log(embeddingBase64Obj);
```
nit
```suggestion
```

RobertCraigie reviewed (changes requested):
Thanks!

IDisposable commented:
This is a great idea! Who doesn't want 1/4 the network bandwidth?

RobertCraigie commented on tests/api-resources/embeddings.test.ts:
```diff
@@ -32,4 +32,38 @@ describe('resource embeddings', () => {
     });
   });
+
+  test.only('create: encoding_format=float should create float32 embeddings', async () => {
```
```suggestion
  test('create: encoding_format=float should create float32 embeddings', async () => {
```

RobertCraigie commented on tests/api-resources/embeddings.test.ts:
```diff
@@ -32,4 +32,38 @@ describe('resource embeddings', () => {
+  });
+
+  test('create: encoding_format=default should create float32 embeddings', async () => {
+    const responsePromise = client.embeddings.create({
+      input: 'The quick brown fox jumped over the lazy dog',
+      model: 'text-embedding-3-small',
+    });
+    const response = await responsePromise;
```
nit:
```suggestion
    const response = await client.embeddings.create({
      input: 'The quick brown fox jumped over the lazy dog',
      model: 'text-embedding-3-small',
    });
```

RobertCraigie commented on tests/api-resources/embeddings.test.ts:
```diff
@@ -32,4 +32,38 @@ describe('resource embeddings', () => {
+  });
+
+  test('create: encoding_format=base64 should create float32 embeddings', async () => {
+    const responsePromise = client.embeddings.create({
+      input: 'The quick brown fox jumped over the lazy dog',
+      model: 'text-embedding-3-small',
+      encoding_format: 'base64',
+    });
+    const response = await responsePromise;
```
nit:
```suggestion
    const response = await client.embeddings.create({
      input: 'The quick brown fox jumped over the lazy dog',
      model: 'text-embedding-3-small',
      encoding_format: 'base64',
    });
```

RobertCraigie commented on tests/api-resources/embeddings.test.ts:
```diff
@@ -32,4 +32,38 @@ describe('resource embeddings', () => {
   });
+
+  test.only('create: encoding_format=float should create float32 embeddings', async () => {
+    const responsePromise = client.embeddings.create({
+      input: 'The quick brown fox jumped over the lazy dog',
+      model: 'text-embedding-3-small',
+    });
+    const response = await responsePromise;
```
nit:
```suggestion
    const response = await client.embeddings.create({
      input: 'The quick brown fox jumped over the lazy dog',
      model: 'text-embedding-3-small',
    });
```

RobertCraigie commented on src/resources/embeddings.ts:
```diff
@@ -9,9 +9,53 @@ export class Embeddings extends APIResource {
+
+    if (hasUserProvidedEncodingFormat) {
+      Core.debug('Request', 'User defined encoding_format:', body.encoding_format);
+    } else {
+      // No encoding_format specified, defaulting to base64 for performance reasons
+      // See https://github.com/openai/openai-node/pull/1312
+      encoding_format = 'base64';
+    }
```
nit: this `else` block is redundant because you already default to `base64` on line 18, comment is useful though so can you move the comment to the assignment above?

manekinekko replied:
fixed.

RobertCraigie commented on src/resources/embeddings.ts:
```diff
@@ -9,9 +9,53 @@ export class Embeddings extends APIResource {
+    // and we defaulted to base64 for performance reasons
+    // we are sure then that the response is base64 encoded, let's decode it
+    // the returned result will be a float32 array since this is OpenAI API's default encoding
+    Core.debug('response', `User requested encoding_format=${encoding_format || 'default'}`);
```
this log is redundant and misleading as there is no user requested format as we're hitting the case where they did not provide a format
```suggestion
```

RobertCraigie commented on src/resources/embeddings.ts:
```diff
@@ -9,9 +9,53 @@ export class Embeddings extends APIResource {
+
+    // if the user specified an encoding_format, return the response as-is
+    if (hasUserProvidedEncodingFormat) {
+      return response as Core.APIPromise<CreateEmbeddingResponse>;
```
nit: do the cast to `Core.APIPromise` on the `response` variable instead of here

manekinekko replied:
fixed.

RobertCraigie reviewed (changes requested):
Sorry for the delayed review, this looks good! Some minor comments and you left a `test.only` change in.

Will merge once comments have been addressed.

manekinekko commented:
> Sorry for the delayed review, this looks good! Some minor comments and you left a `test.only` change in.
> 
> Will merge once comments have been addressed.

@RobertCraigie no worries about the delay. Thank you for reviews. I addressed your suggestions.

manekinekko commented:
@RobertCraigie PR is now ready for review. Thank you.

RobertCraigie commented on src/core.ts:
```diff
@@ -1287,6 +1287,27 @@ export const toBase64 = (str: string | null | undefined): string => {
+export const toFloat32Array = (base64Str: string): Array<number> => {
+  if (typeof Buffer !== 'undefined') {
+    // for Node.js environment
+    return Array.from(new Float32Array(Buffer.from(base64Str, 'base64').buffer));
```
curious if you've benchmarked how much of a difference just returning the `Float32Array` directly would have?

if it's a big difference we should probably have an opt-in flag to just do that. (doesn't block this PR)

RIscRIpt replied:
This implementation seems to produce wrong results, see #1448

manekinekko replied:
Awesome! Thank you @RIscRIpt for looking into this.

RobertCraigie reviewed (approved):
Thanks! Sorry again for the delay.

I pushed a commit removing some debug logs as our debug logging system is not particularly great right now so they'd be too verbose IMO. (logging will be fixed in the next major version)
